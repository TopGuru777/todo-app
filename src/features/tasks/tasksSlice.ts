import { createSlice, PayloadAction } from '@reduxjs/toolkit';

type TaskStatus = 'Pending' | 'In Progress' | 'Completed';

export interface Task {
  id: string;
  title: string;
  description: string;
  status: TaskStatus;
  dueDate: string;
}

interface TasksState {
  tasks: Task[];
}

const initialState: TasksState = {
  tasks: [],
};

const tasksSlice = createSlice({
  name: 'tasks',
  initialState,
  reducers: {
    addTask: (state, action: PayloadAction<Task>) => {
      state.tasks.push(action.payload);
    },
    editTask: (state, action: PayloadAction<Task>) => {
      const index = state.tasks.findIndex(task => task.id === action.payload.id);
      if (index !== -1) {
        state.tasks[index] = action.payload;
      }
    },
    deleteTask: (state, action: PayloadAction<string>) => {
      state.tasks = state.tasks.filter(task => task.id !== action.payload);
    },
    updateTasks: (state, action: PayloadAction<Task[]>) => {
      state.tasks = action.payload;
    },
  },
});

export const { addTask, editTask, deleteTask, updateTasks } = tasksSlice.actions;
export default tasksSlice.reducer;