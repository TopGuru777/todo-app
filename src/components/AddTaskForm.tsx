import React, { useState, useCallback } from 'react';
import { useDispatch } from 'react-redux';
import { addTask } from '../features/tasks/tasksSlice';
import { v4 as uuidv4 } from 'uuid';

const AddTaskForm: React.FC = () => {
  const dispatch = useDispatch();
  const [title, setTitle] = useState('');
  const [description, setDescription] = useState('');
  const [dueDate, setDueDate] = useState('');

  const handleSubmit = useCallback((e: React.FormEvent) => {
    e.preventDefault();
    if (!title || !dueDate) return;

    dispatch(addTask({
      id: uuidv4(),
      title,
      description,
      status: 'Pending',
      dueDate,
    }));
    setTitle('');
    setDescription('');
    setDueDate('');
  }, [dispatch, title, description, dueDate]);

  return (
    <form onSubmit={handleSubmit} className="bg-white shadow-md rounded px-8 pt-6 pb-8 mb-4">
      <h2 className="text-xl font-bold mb-4 text-center">Add New Task</h2>
      <div className="mb-4">
        <input
          type="text"
          value={title}
          placeholder="Task title"
          onChange={(e) => setTitle(e.target.value)}
          required
          className="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
        />
      </div>
      <div className="mb-4">
        <textarea
          value={description}
          placeholder="Task description"
          onChange={(e) => setDescription(e.target.value)}
          className="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
        />
      </div>
      <div className="mb-4">
        <input
          type="date"
          value={dueDate}
          data-testid="due-date-input"
          onChange={(e) => setDueDate(e.target.value)}
          required
          className="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
        />
      </div>
      <div className="flex items-center justify-between">
        <button 
          type="submit" 
          className="bg-blue-600 hover:bg-blue-700 text-white font-semibold py-2 px-4 rounded-lg shadow-md transition duration-300"
        >
          Add Task
        </button>
      </div>
    </form>
  );
};

export default AddTaskForm;