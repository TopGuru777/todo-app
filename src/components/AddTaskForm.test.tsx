import { render, fireEvent, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import AddTaskForm from './AddTaskForm';
import { addTask } from '../features/tasks/tasksSlice';

const mockStore = configureStore([]);
jest.mock('uuid', () => ({ v4: () => '1234' }));
jest.mock('../features/tasks/tasksSlice', () => ({
  addTask: jest.fn(),
}));

describe('AddTaskForm', () => {
  let store: ReturnType<typeof mockStore>;

  beforeEach(() => {
    store = mockStore({});
    store.dispatch = jest.fn();
  });

  test('submits form and dispatches addTask action', () => {
    const { getByPlaceholderText, getByText } = render(
      <Provider store={store}>
        <AddTaskForm />
      </Provider>
    );

    fireEvent.change(getByPlaceholderText(/Task title/i), { target: { value: 'New Task' } });
    fireEvent.change(getByPlaceholderText(/Task description/i), { target: { value: 'New Description' } });
    fireEvent.change(screen.getByTestId('due-date-input'), { target: { value: '2023-11-01' } });

    fireEvent.click(getByText(/Add Task/i));

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(addTask({
      id: '1234',
      title: 'New Task',
      description: 'New Description',
      status: 'Pending',
      dueDate: '2023-11-01',
    }));
  });

  test('matches snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <AddTaskForm />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});