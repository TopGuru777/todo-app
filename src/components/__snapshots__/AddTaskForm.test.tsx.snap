// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTaskForm matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="bg-white shadow-md rounded px-8 pt-6 pb-8 mb-4"
  >
    <h2
      class="text-xl font-bold mb-4 text-center"
    >
      Add New Task
    </h2>
    <div
      class="mb-4"
    >
      <input
        class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
        placeholder="Task title"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-4"
    >
      <textarea
        class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
        placeholder="Task description"
      />
    </div>
    <div
      class="mb-4"
    >
      <input
        class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
        data-testid="due-date-input"
        required=""
        type="date"
        value=""
      />
    </div>
    <div
      class="flex items-center justify-between"
    >
      <button
        class="bg-blue-600 hover:bg-blue-700 text-white font-semibold py-2 px-4 rounded-lg shadow-md transition duration-300"
        type="submit"
      >
        Add Task
      </button>
    </div>
  </form>
</DocumentFragment>
`;
